Comparison of size_t and int is sometimes broken

The current logic does not work if the arguments to `assert_equal`/`assert_not_equal` are deduced to be by reference rather than by value. In fact, the existing `size_t_and_int.cpp` test currently fails with GCC and `-Wall` and with Clang and `-Wextra`:
```
In file included from size_t_and_int.cpp:1:
unit_test_framework.h: In instantiation of 'static bool safe_equals<First, Second, <template-parameter-1-3> >::not_equals(const First&, const Second&) [with First = long unsigned int&; Second = int&; <template-parameter-1-3> = void]':
unit_test_framework.h:325:47:   required from 'void assert_not_equal(First&&, Second&&, int) [with First = long unsigned int&; Second = int&]'
size_t_and_int.cpp:7:3:   required from here
unit_test_framework.h:292:22: warning: comparison of integer expressions of different signedness: 'long unsigned int' and 'int' [-Wsign-compare]
  292 |         return first != second;
      |                ~~~~~~^~~~~~~~~
unit_test_framework.h: In instantiation of 'static bool safe_equals<First, Second, <template-parameter-1-3> >::equals(const First&, const Second&) [with First = int&; Second = long unsigned int&; <template-parameter-1-3> = void]':
unit_test_framework.h:313:43:   required from 'void assert_equal(First&&, Second&&, int) [with First = int&; Second = long unsigned int&]'
size_t_and_int.cpp:8:3:   required from here
unit_test_framework.h:289:22: warning: comparison of integer expressions of different signedness: 'int' and 'long unsigned int' [-Wsign-compare]
  289 |         return first == second;
      |                ~~~~~~^~~~~~~~~
```

The logic needs to be reworked to handle both value and reference parameters.

Also, the comparisons should check whether the `int` argument is nonnegative.

--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,6 @@
 CXX = g++
 CPPFLAGS = -MMD -MP -I.
-CXXFLAGS = -Wall -Wextra -pedantic --std=c++11
+CXXFLAGS = -Wall -Wextra -Werror -pedantic --std=c++11
 
 test_dir = test
 
@@ -8,6 +8,7 @@ test_exes = $(test_dir)/test_special_asserts.exe \
 			$(test_dir)/end_to_end_test.exe \
 			$(test_dir)/test_early_exit.exe \
 			$(test_dir)/sequence_equal.exe \
+			$(test_dir)/size_t_and_int.exe \
 			$(test_dir)/string_equal.exe
 
 end_to_end_corrects = $(wildcard $(test_dir)/*.correct.txt)
@@ -34,6 +35,7 @@ test: $(test_exes) $(test_dir)/end_to_end_test.exe
 	$(call run_and_diff,$(test_dir)/end_to_end_test.exe should_pass -q should_fail,quiet_run_subset_of_tests)
 	$(call run_and_diff,$(test_dir)/sequence_equal.exe,sequence_equal)
 	$(call run_and_diff,$(test_dir)/string_equal.exe,string_equal)
+	./$(test_dir)/size_t_and_int.exe
 
 %.o: %.cpp
 	$(CXX) $(CPPFLAGS) $(CXXFLAGS) $< -c -o $@

--- a/test/end_to_end_test.cpp
+++ b/test/end_to_end_test.cpp
@@ -22,4 +22,4 @@ TEST(assert_almost_equal_precision_infinite) {
     ASSERT_ALMOST_EQUAL(.2 + .1, .3, 0);
 }
 
-TEST_MAIN();
+TEST_MAIN()

--- a/test/size_t_and_int.cpp
+++ b/test/size_t_and_int.cpp
@@ -5,6 +5,8 @@ TEST(size_t_and_int) {
   int i1 = -1;
   int i2 = 3;
   ASSERT_NOT_EQUAL(s, i1);
+  ASSERT_NOT_EQUAL(i1, s);
+  ASSERT_EQUAL(s, i2);
   ASSERT_EQUAL(i2, s);
 }
 

--- a/test/test_name_conflict_file_with_tests.cpp
+++ b/test/test_name_conflict_file_with_tests.cpp
@@ -4,4 +4,4 @@
 // test_name_conflict_file_with_non_static_func.cpp.
 TEST(my_test) {}
 
-TEST_MAIN();
+TEST_MAIN()

--- a/unit_test_framework.h
+++ b/unit_test_framework.h
@@ -281,30 +281,43 @@ struct is_equality_comparable<First, Second,
                               enable_if_equality_comparable<First, Second>>
     : std::true_type {};
 
+// Overloads for equality comparisons.
+template <typename First, typename Second>
+bool safe_equals_helper(const First& first, const Second& second) {
+  return first == second;
+}
+
+template <typename First, typename Second>
+bool safe_not_equals_helper(const First& first, const Second& second) {
+  return first != second;
+}
+
+// Allow size_t to correctly be compared to int.
+bool safe_equals_helper(std::size_t first, int second) {
+  return second >= 0 && static_cast<long long>(first) == second;
+}
+
+bool safe_equals_helper(int first, std::size_t second) {
+  return first >= 0 && first == static_cast<long long>(second);
+}
+
+bool safe_not_equals_helper(std::size_t first, int second) {
+  return second < 0 || static_cast<long long>(first) != second;
+}
+
+bool safe_not_equals_helper(int first, std::size_t second) {
+  return first < 0 || first != static_cast<long long>(second);
+}
+
 template <typename First, typename Second, typename = void>
 struct safe_equals {
     static_assert(is_equality_comparable<First, Second>::value,
                   "types cannot be compared with == and !=");
     static bool equals(const First& first, const Second& second) {
-        return first == second;
+        return safe_equals_helper(first, second);
     }
     static bool not_equals(const First& first, const Second& second) {
-        return first != second;
-    }
-};
-
-// Specializations to allow size_t to be compared to int.
-template <>
-struct safe_equals<std::size_t, int> {
-    static bool equals(std::size_t first, int second) {
-        return static_cast<long long>(first) == second;
-    }
-};
-
-template <>
-struct safe_equals<int, std::size_t> {
-    static bool equals(int first, std::size_t second) {
-        return first == static_cast<long long>(second);
+        return safe_not_equals_helper(first, second);
     }
 };
 

